Feature/redeem-form: implement confirmation alert

mateuszRybczonek opened the pull request:
This PR:
- adds a Confirmation Modal that is shown when the redemption is successfully initiated
<img width="751" alt="Screenshot 2023-04-28 at 13 45 00" src="https://user-images.githubusercontent.com/17565389/235139132-3fb62a49-229d-4973-8db7-1a9f4d50d946.png">

- hooks up API calls to "Confirm & Redeem" button that makes calls using zome_call:

1. `get_all_reserve_accounts_details` - to get the reserve details used in the second call
2. `redemption` - to initiate redemption 

- redirects to earnings when the redemption is successfully initiated

- checks redeemable amount just before making a call for redemption and returns to step one with an error message if the requested amount exceeds redeemable amount
<img width="1496" alt="Screenshot 2023-04-28 at 15 20 51" src="https://user-images.githubusercontent.com/17565389/235159604-ecb9a41c-faad-49bd-9306-103b5a1eb9c6.png">

- if redeem API call fails it returns to step one and shows the error banner
<img width="1501" alt="Screenshot 2023-04-28 at 15 19 43" src="https://user-images.githubusercontent.com/17565389/235159742-5f73c1a1-8dc3-4285-86e2-06f8162d3261.png">

mateuszRybczonek committed:
feat(redeem-form): implement confirmation alert and connect to mocked API

mateuszRybczonek committed:
feat(redeem-form): update ui-common-library

mateuszRybczonek committed:
Merge branch 'develop' into feature/hook-up-redemption-endpoint

mateuszRybczonek committed:
feat(redeem-form): remove duplicated css code

mateuszRybczonek committed:
feat(redeem-form): error handling

- check redeemable before calling redeem endpoint, redirect to step 1 and show error if requested value is bigger
- if redeem API call fails return to step 1 and show banner

pegaltier commented:
@mateuszRybczonek @marycamacho @artbrock @zippy @alastairong  My suggestion is to use the ticker XHF instead of HF indeed ‘supranational’ currency with no official country code and therefore must begin with an ‘X’ as per the required standards for business and banking globally by [ISO_4217]. See: https://blog.nano.org/say-hello-to-xno-76568e1b76a0

JettTech commented on mock-hpos-api/defaultResponse.js:
```diff
@@ -190,6 +190,40 @@ const getMyProfile = {
 }
 
+const getAllReserveAccountsDetails = {
+  hash: 'QmYXJ5Y2FyZG9uLmNvbS9ob2xvZnVsb2c=',
```
```suggestion
  reserve_id: 'QmYXJ5Y2FyZG9uLmNvbS9ob2xvZnVsb2c=',
```

JettTech replied:
This var has been renamed to `reserve_id`.

JettTech commented on mock-hpos-api/defaultResponse.js:
```diff
@@ -190,6 +190,40 @@ const getMyProfile = {
 
+const getAllReserveAccountsDetails = {
+  hash: 'QmYXJ5Y2FyZG9uLmNvbS9ob2xvZnVsb2c=',
+  pub_key: {
```
The type for `pub_key` here is actually a base64 encoded string with some special "holo hash" properties/prefixes.

The `cryptolib` README and tests provide good examples of  what it will look like: https://github.com/Holo-Host/cryptolib-js/blob/develop/tests/test_codec.js#L19

For simplicity and accuracy, I'd actually suggest just to import the [@holo-host/cryptolib](https://www.npmjs.com/package/@holo-host/cryptolib) package as a dev dependency and use the `AgentId.encode` method itself.
```
import { Codec } from '@holo-host/cryptolib';
Codec.AgentId.encode(pub_key)
```

JettTech commented on mock-hpos-api/defaultResponse.js:
```diff
@@ -190,6 +190,40 @@ const getMyProfile = {
+  direction: '',
+  status: '',
+  note: '',
+  proof_of_service: '',
```
```suggestion
  proof_of_service: {
    redemption: "0x12345678..."
  },
```

This will be the identifying factor of a redemption transaction.

JettTech commented on src/components/earnings/RedeemHoloFuelCard.vue:
```diff
@@ -39,11 +58,34 @@ interface StepTwoProps {
+
+    const transaction: RedemptionTransaction = await redeemHoloFuel({
+      amount: amount.value,
+      note: JSON.stringify({ eth_address: hotAddress })
```
```suggestion
      wallet_address: hotAddress
```

The endpoint implementation has now been updated such that we no longer need to send the `eth_address` in the `note` field.  Instead you'll need to pass it in as `wallet_address` as shown above.

JettTech commented on src/components/earnings/RedeemHoloFuelCard.vue:
```diff
@@ -39,11 +58,34 @@ interface StepTwoProps {
+    await dashboardStore.getEarnings()
+
+    if (Number(redeemableHoloFuel.value) < Number(amount.value)) {
+      partialRedemptionTermsAccepted.value = false
```
Shouldn't we show some sort of user feedback informing the user that they're trying to redeem more than they have available?

mateuszRybczonek replied:
We do, I just added a comment there, basically we redirect user to step one where the error will be shown next to the amount input.

JettTech commented on src/interfaces/HposInterface.ts:
```diff
@@ -639,6 +668,48 @@ export function useHposInterface(): HposInterface {
+        appId: localStorage.getItem(kCoreAppVersionLSKey),
+        roleId: 'holofuel',
+        zomeName: 'transactor',
+        fnName: 'redemption',
```
```suggestion
        fnName: 'redeem',
```
Updating to match changes to implementation.

JettTech commented on src/interfaces/HposInterface.ts:
```diff
@@ -639,6 +668,48 @@ export function useHposInterface(): HposInterface {
+        fnName: 'redemption',
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+        payload: {
+          reserve_id: reserveAccountsDetails.hash,
```
```suggestion
          reserve_id: reserveAccountsDetails.reserve_id,
```

Updating to match changes to implementation.

JettTech commented on src/interfaces/HposInterface.ts:
```diff
@@ -23,9 +23,36 @@ interface HposInterface {
+}
+
+interface ReserveSettingsResponse {
+  hash: string
```
```suggestion
  reserve_id: string
```

Updating to match implementation changes.

JettTech commented on src/interfaces/HposInterface.ts:
```diff
@@ -23,9 +23,36 @@ interface HposInterface {
+  counterparty: string
+  direction: string
+  status: string
+  note?: string // The `eth_hash` will be stored here in Redemption Transactions
```
```suggestion
  note?: string
  proof_of_service?:  {
    redemption: string, // The `eth_hash` (ie: wallet_address) will be stored here in Redemption Transactions
  }, 
```

Updating to match implementation changes.

JettTech commented on src/interfaces/HposInterface.ts:
```diff
@@ -23,9 +23,36 @@ interface HposInterface {
 
+interface RedeemHoloFuelPayload {
+  amount: string
+  note: string
```
```suggestion
  note: string,
  wallet_address: string
```

JettTech replied:
^^ Updating to match implementation changes.

JettTech commented on src/components/modals/RedemptionInitiatedModal.vue:
```diff
@@ -0,0 +1,153 @@
+  const rawRedemptionAmount: string = modalProps.value.redemptionAmount
+  return rawRedemptionAmount && Number(rawRedemptionAmount)
+    ? formatCurrency(Number(rawRedemptionAmount))
+    : 0
```
Should we ever just return 0?  If `rawRedemptionAmount` doesn't exist, shouldn't an error be (...or already have been) thrown?

mateuszRybczonek replied:
True, thanks

JettTech commented on src/components/modals/RedemptionInitiatedModal.vue:
```diff
@@ -0,0 +1,153 @@
+
+const formattedHfAmount = computed((): number | string => {
+  const rawHfAmount: string = modalProps.value.hfAmount
+  return rawHfAmount && Number(rawHfAmount) ? formatCurrency(Number(rawHfAmount)) : 0
```
Same as above ...should we ever just return 0? If `rawHfAmount` doesn't exist, shouldn't an error be (...or already have been) thrown?

mateuszRybczonek replied:
Correct

JettTech commented on src/pages/RedeemHoloFuelPage.vue:
```diff
@@ -29,13 +34,25 @@ async function getRedeemableHoloFuel(): Promise<void> {
+    date: transaction.created_date,
+    hfAmount: transaction.amount,
+    requestId: transaction.id,
+    currency: 'HOT',
```
Can we add a quick note explaining that this is only hard coded until we add the call to support other Reserve types (functionality already exists in back-end).

JettTech commented on src/components/modals/RedemptionInitiatedModal.vue:
```diff
@@ -0,0 +1,153 @@
+  {
+    label: `${t('redeem_holofuel.redemption_amount')}:`,
+    value: `${formattedRedemptionAmount.value} HOT`,
+    tipMessage: t('redemption_history.transaction_price', { hf: 1, hot: 1 })
```
It appears there are now two references to `redemption_history`. Can we collapse one into the other as  a `header` or `title`?  If not, we should probably rename one.

JettTech replied:
...also why does `transaction_price` live on `redemption_history`?  It will be something that is needed and accessible *prior* to making a request to `redeem` holofuel, and therefore isn't necessarily related to the `redemption_history`.

JettTech reviewed (changes requested):
Overall looks good. I've left a few small comments and suggestions for required implementation updates to match the changes to the backend. 

I'd like to do one more quick pass to make sure all is correct after these updates.  Let me  know if you have any Qs.

mateuszRybczonek committed:
feat(redeem-form): address review comments

- do not return 0 on computed props as the value should always exist
- extract common translation
- update API call and response acc. to latest BE implementation

mateuszRybczonek committed:
feat(redeem-form): reset step two checkbox if redemption fails

JettTech commented on src/components/earnings/RedeemHoloFuelCard.vue:
```diff
@@ -67,6 +69,8 @@ async function handleSubmit(): Promise<void> {
 
     if (Number(redeemableHoloFuel.value) < Number(amount.value)) {
+      // Reset the checkbox and redirect to step 1 where
+      // user can see the error and enter a new amount
```
Love this contextual note addition.

JettTech reviewed (approved):
Nice updates.  Lgtm. :)
